### PR TITLE
[RFC] introduce common periph led driver

### PR DESCRIPTION
**Pull request by smlng**
this PR introduces a simplified handling of on-board LEDs by fully utilizing RIOTs hardware independent gpio driver. That way per board redefinition of led macros such as `LEDX_ON`, `LEDX_OFF`, and `LEDX_TOGGLE` will be obsolete, it simple requires to define the respective gpio pin for each LED as it is already done. 

As an example this PR shows this feature on board `remote-revb`. However, this is foremost a Request-For-Comments and hence Work-In-Progress.

### CODE REVIEW

**Comment by miri64** [comment]
What is the reasoning for the change? What is the overall size benefit/drawback?

File: drivers/include/periph/led.h [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+/* assuming LED7_OFF and LED7_TOGGLE are undefined too */
+#define LED7_ON                 gpio_set(LED7_PIN)
+#define LED7_OFF                gpio_clear(LED7_PIN)
+#define LED7_TOGGLE             gpio_toggle(LED7_PIN)
+#endif /* LED7_ON */
+#ifndef LED7_INIT
+#define LED7_INIT               do { gpio_init(LED7_PIN, GPIO_OUT); LED7_OFF; } while(0)
+#endif /* LED7_INIT */
+#endif /* LED7_PIN */
+/** @} */
```
Comment (kYc0o) [inline drivers/include/periph/led.h]: Maybe instead of creating a define per "possibly" LEDs present in a board (what if I have 8 LEDs? This won't work), you can propose a definition similar to UART, SPI or (future) I2C, on which you have a general structure and is the board which defines how many "objects" of this kind it has, this way you'll avoid the need to define everything possible.

However, I doubt if such a solution would have a negative impact on code size.

File: drivers/periph_common/led.c [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#endif /* LED5_PIN */
+
+#ifdef LED6_PIN
+    LED6_INIT;
+#endif /* LED6_PIN */
+
+#ifdef LED7_PIN
+    LED7_INIT;
+#endif /* LED7_PIN */
+
```
Comment (kYc0o) [inline drivers/periph_common/led.c]: Here too, same comment as above.

**Comment by smlng** [comment]
the reasoning behind this is that currently each board redefines macros `LED0_[ON|OFF|TOGGLE]` which is unnecessary as all boards define on-board LEDs using the `GPIO_PIN` macro of the common gpio API. Further, all of these macros  (e.g. `LED0_ON`) use CPU specific gpio handling which is already implemented by the CPU specific low lever gpio driver `cpu/<cpu_name>/periph/gpio.c` - compare for instance [1] and [2]

[1]: https://github.com/RIOT-OS/RIOT/blob/master/boards/remote-revb/include/board.h#L46
[2]: https://github.com/RIOT-OS/RIOT/blob/master/cpu/cc2538/periph/gpio.c#L151

File: drivers/include/periph/led.h (hunk 1 above)
Comment (smlng) [inline drivers/include/periph/led.h]: I thought about this, too ... will try this later on. For now this PR is mostly about avoiding redefinition per board for common function such as `LED[0-7]_ON`

**Comment by smlng** [comment]
btw. @miri64 regarding code size, using `tests/led` this already safes 280B 😮 

master
```
text	   data	    bss	    dec	    hex	filename
9104	    136	   2860	  12100	   2f44	/RIOT/tests/leds/bin/remote-revb/tests_leds.elf
```
this PR
```
text	   data	    bss	    dec	    hex	filename
8824	    136	   2860	  11820	   2e2c	/RIOT/tests/leds/bin/remote-revb/tests_leds.elf
```

File: drivers/include/periph/led.h (hunk 1 above)
Comment (smlng) [inline drivers/include/periph/led.h]: btw. 8 LEDs would still work (0-7, it is). However, supporting arbitrary numbers of LEDs is difficult while using macros

**Comment by gebart** [comment]
Why not just use the normal GPIO API and create macros as aliases for the GPIO pins?
Instead of LED0_ON, LED0_ON, LED0_TOGGLE you will only need LED_0 and call gpio_init(LED_0), gpio_set(LED_0) etc

**Comment by smlng** [comment]
@gebart that's already and still possible - though not with `LED_0` but `LED0_PIN`, which is by the way what this PR uses, eg. `LED0_ON` is defined as `gpio_set(LED0_PIN)`. However, we have and use the util macros `LED0_ON|OFF|TOGGLE` throughout the code, so IMHO it would be bad to remove these short hand macros.

**Comment by smlng** [comment]
and as I already said, primary goal of this PR is to remove redundant redefinition of LED macros and that way also reduce stuff defined in `board.h` and or `periph_conf.h` on a per board basis.

**Comment by smlng** [comment]
I ported this to `samr21-xpro` and `pba-d-01-kw2x` where it introduces an overhead of >100B in both cases. 

Still I think its worth discussing and reconsidering current handling of on-board leds - but not the way I proposed here, I'll try something different, too.

Will close this soonish as memo, just keep it open if someone wants to had her/his 2 cents

**Comment by haukepetersen** [comment]
I agree that we should revisit this topic, its been a while since we had this on our list (hello @OlegHahm :-) ).

To recab: the reason for having the LED macros was/is simply to be able to control selected pins (i.e. those that have LEDs connected to them) with the least amount of overhead, so they can used for timing analysis. I personally don't think this use is not very pressing (anymore), as most timing analysis are carried out using hardware timers and/or other profiling solutions.

Also I think the usefulness of the LED macros for debugging has quite degraded over time, as we have access to decent debuggers on most platforms, that make the use of LED macros for debugging obsolete on these.

Bottom line: I think we can deprecate the existing LED macros and move to something easier maintainable.

I do however not like the proposed solution:
- the LED handling does not belong into `periph`! LEDs do not have anything to do with MCU peripherals!
- we can simplify the LED handling even more, just played around a bit and came up with #7350

**Comment by smlng** [comment]
@haukepetersen your #7350 is basically what I had in mind with:

> but not the way I proposed here, I'll try something different,

so, now its really time to close this and move the discussion to your PR ...
